synapse/handlers/stats.py

```diff
@@ -52,4 +59,314 @@ def __init__(self, hs):
+                logger.debug(
+                    "Ignoring: %s as earlier than this room's initial ingestion event",
+                    event_id,
+                )
+                continue
+
+            if event_id is None and prev_event_id is None:
+                # Errr...
+                continue
+
+            event_content = {}
+
+            if event_id is not None:
+                event = yield self.store.get_event(event_id, allow_none=True)
+                if event:
+                    event_content = event.content or {}
+
+            # We use stream_pos here rather than fetch by event_id as event_id
+            # may be None
+            now = yield self.store.get_received_ts_by_stream_pos(stream_pos)
+            now = int(now)
+
+            room_stats_delta = {}
+            room_stats_complete = False
+
+            if prev_event_id is None:
+                # this state event doesn't overwrite another,
+                # so it is a new effective/current state event
+                room_stats_delta["current_state_events"] = 1
+
+            if typ == EventTypes.Member:
+                # we could use _get_key_change here but it's a bit inefficient
+                # given we're not testing for a specific result; might as well
+                # just grab the prev_membership and membership strings and
+                # compare them.
+                # We take None rather than leave as a previous membership
+                # in the absence of a previous event because we do not want to
+                # reduce the leave count when a new-to-the-room user joins.
+                prev_membership = None
+                if prev_event_id is not None:
+                    prev_event = yield self.store.get_event(
+                        prev_event_id, allow_none=True
+                    )
+                    if prev_event:
+                        prev_event_content = prev_event.content
+                        prev_membership = prev_event_content.get(
+                            "membership", Membership.LEAVE
+                        )
+
+                membership = event_content.get("membership", Membership.LEAVE)
+
+                if prev_membership is None:
+                    logger.debug("No previous membership for this user.")
+                elif membership == prev_membership:
+                    pass  # noop
+                elif prev_membership == Membership.JOIN:
+                    room_stats_delta["joined_members"] = -1
+                elif prev_membership == Membership.INVITE:
+                    room_stats_delta["invited_members"] = -1
+                elif prev_membership == Membership.LEAVE:
+                    room_stats_delta["left_members"] = -1
+                elif prev_membership == Membership.BAN:
+                    room_stats_delta["banned_members"] = -1
+                else:
+                    err = "%s is not a valid prev_membership" % (repr(prev_membership),)
```

Comment:
`repr(..)` here is the same as using `%r` (`%r` calls `repr(..)` on the arg and `%s` call `str(..)` on the arg).

Also, lets just have this in the `logger.error` below so we don't have to manually do percent formatting?

```python
logger.error(
    "%r is not a valid prev_membership",
    prev_membership,
)
```